activemq-ra/src/main/java/org/apache/activemq/ra/ServerSessionImpl.java

```diff
@@ -64,8 +65,8 @@ public class ServerSessionImpl implements ServerSession, InboundContext, Work, D
     private MessageProducer messageProducer;
     private final ServerSessionPoolImpl pool;
 
-    private Object runControlMutex = new Object();
-    private boolean runningFlag;
```

Comment:
I guess the problem is here as the `runningFlag` is not thread safe (and we can have several session), right ?

Reply:
This is a bit of a trick, because this avoids the lock but does not fix yet the initial issue in my opinion. There is an underlying bug revealed by this deadlock allowing another thread to be using this session to dispatch unconsumed messages from a stale session (removeFromPool).

Trying to properly understand it, so I can push a more complete fix

Reply:
So the trick can happen when there is no more session in the pool, then ActiveMQ pulls an active session leading to a deadlock as I have observed. With the AtomicInterger, the integrity is preserved without any locks.